### PR TITLE
Symlink CASEDIR and set values to relative path

### PR DESCRIPTION
**Pull request by Amrysliu**
1. symlink test repository from CASEDIR to POOLDIR (using CASEDIR's 
basename as the destdir name).
2. set the CASEDIR and PRODUCTDIR as relative values.
3. when uing `openqa-clone-custom-git-refspec`, set the NEEDLES_DIR to
`needles`, because code in openQA worker will symlink the needles
directory to it.

Verify run:
The source job run on a remote worker: http://10.67.19.103/tests/413
The clone job run on a local worker: http://10.67.19.103/tests/414#settings
A simple case use relative path as CASEDIR and PRODUCTDIR: http://10.67.19.103/tests/408/file/vars.json

Relate: https://progress.opensuse.org/issues/80372

### CODE REVIEW

**Comment by Amrysliu** [comment]
I am not sure if I am in the right way according to suggestions in https://progress.opensuse.org/issues/80372. I am also confused about the `Then in the openQA worker we create the symlink as mentioned above and do not need to specify CASEDIR at all`. Why we need this symlink? In general, the `DEFAULT_CASEDIR` should exist? Could we access it directly?

**Comment by kalikiana** [comment]
> I am not sure if I am in the right way according to suggestions in https://progress.opensuse.org/issues/80372. I am also confused about the `Then in the openQA worker we create the symlink as mentioned above and do not need to specify CASEDIR at all`. Why we need this symlink? In general, the `DEFAULT_CASEDIR` should exist? Could we access it directly?
> 
> The related code in os-autoinst: [os-autoinst/os-autoinst#1577](https://github.com/os-autoinst/os-autoinst/pull/1577)

The way I understand the suggestion is this. We have multiple possible directories and may not know what the correct one is in the beginning. Instead of adding another variable to check, we could have a symlink to the real dir. So the path never changes. Even if the dir is different.

**Comment by Amrysliu** [comment]
> > I am not sure if I am in the right way according to suggestions in https://progress.opensuse.org/issues/80372. I am also confused about the `Then in the openQA worker we create the symlink as mentioned above and do not need to specify CASEDIR at all`. Why we need this symlink? In general, the `DEFAULT_CASEDIR` should exist? Could we access it directly?
> > The related code in os-autoinst: [os-autoinst/os-autoinst#1577](https://github.com/os-autoinst/os-autoinst/pull/1577)
> 
> The way I understand the suggestion is this. We have multiple possible directories and may not know what the correct one is in the beginning. Instead of adding another variable to check, we could have a symlink to the real dir. So the path never changes. Even if the dir is different.

But if we do not add another variable, how could we know the real dir?  Since we know the real dir, why not just access it instead of making the symlink?

**Comment by Martchus** [comment]
I also don't understand the symlink suggestion (and hence haven't commented on your PRs yet). I suppose @okurz who suggested it should leave a comment.

**Comment by okurz** [comment]
ok, so some thoughts:
* My idea about symlink was that the worker uses its "pool dir" as current working directory and we already use the "pool dir" for the git-clone use case so same as we link some assets from "cache" to "pool" we could link the test case directory from "cache" to "pool" and the worker as well as isotovideo would be able to use a relative lookup, same as we already use for the git clone case. Then casedir could simply be something like "os-autoinst-distri-opensuse" (absolute path for example /var/lib/openqa/pool/1/os-autoinst-distri-opensuse) which could be a symlink that points to e.g. /var/lib/openqa/cache/openqa.suse.de/tests/sle . As it seems openQA does not really *know* that this is a git checkout of os-autoinst-distri-opensuse it's ok if we just call a symlink with the name "sle" as well.
* My idea was to prevent the use of any non-portable absolute paths so that anyone can call isotovideo in a similar way locally or at least properly clone to any other worker that can work with the test definition without needing to replicate the path structure, regardless if using "caching" where currently we have something like CASEDIR=/var/lib/openqa/cache/openqa.suse.de/tests/sle or "non-caching" where we have something like CASEDIR=/var/lib/openqa/share/tests/opensuse
* Be aware of the different directory definitions and their meanings in https://github.com/os-autoinst/os-autoinst/blob/master/doc/backend_vars.asciidoc
* Here no tests fail likely because os-autoinst uses the internal test distribution from https://github.com/os-autoinst/os-autoinst/tree/master/t/data/tests so only CASEDIR is necessary and PRODUCTDIR is not necessary. To prevent a regression for test distributions that rely on PRODUCTDIR one should do appropriate tests as well or extend the os-autoinst test distribution accordingly

**Comment by Amrysliu** [comment]
@okurz Thanks for more details.
please help me to confirm that:
1. we don't need `PRODUCTDIR`, but we need to do back compatibility.
2. we could always do `symlink`, such as link `/var/lib/openqa/share/tests/sle` to `/var/lib/openqa/pool/1/sle`, then isotovideo can use the relative path `sle`. 
3. If the above 2 is right, we need another parameter to record the relative pathname or using the `$DISTRI` and `$VERSION`? Because sometimes the `CASEDIR` is a GitHub path.

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 1]
```diff
@@ -323,10 +334,13 @@ sub engine_workit {
         return $error if $error;
     }
 
-    $vars{ASSETDIR}   //= OpenQA::Utils::assetdir();
-    $vars{CASEDIR}    //= OpenQA::Utils::testcasedir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
-    $vars{PRODUCTDIR} //= OpenQA::Utils::productdir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
+    my $casedir   = path(OpenQA::Utils::testcasedir($vars{DISTRI}, $vars{VERSION}, $shared_cache));
+    my $test_repo = $casedir->basename;
+    $vars{ASSETDIR} //= OpenQA::Utils::assetdir();
+    $vars{TESTREPO} = $test_repo;
```
Comment (okurz) [inline lib/OpenQA/Worker/Engines/isotovideo.pm:340]: Please consider https://en.wikipedia.org/wiki/Test-driven_development as this can also help reviewers understand your intentions much better.

I assume you introduced the "TESTREPO" due to the problematic needle repo which can be outside CASEDIR in case of the "clone-custom-git-refspec" approach. I struggle with the name and meaning of "TESTREPO" though as for me it has just the same meaning as "CASEDIR", which is short for "test case directory".

**Review by okurz — CHANGES_REQUESTED** [review]

**Comment by okurz** [comment]
> @okurz Thanks for more details.
> please help me to confirm that:
> 
>     1. we don't need `PRODUCTDIR`, but we need to do back compatibility.

Well, we still *need it* but we do not necessarily need to set it in openQA when we handle it in os-autoinst

>     2. we could always do `symlink`, such as link `/var/lib/openqa/share/tests/sle` to `/var/lib/openqa/pool/1/sle`, then isotovideo can use the relative path `sle`.

yes

>     3. If the above 2 is right, we need another parameter to record the relative pathname or using the `$DISTRI` and `$VERSION`? Because sometimes the `CASEDIR` is a GitHub path.

We should be able to use CASEDIR for that. In the case of the openQA worker having the distribution available that should be linked with symlink into the pool dir regardless if the test distribution comes from "caching" or "non-caching". And for "CASEDIR is a github path" then openQA does not need to do anything as git is cloned by isotovideo which cares about setting CASEDIR internally then

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 1 above)
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:340]: yes, 'TESTREPO' is used to find the needles. My intention about this parameter is `test repository`.

**Comment by okurz** [comment]
Yes, I understand. But this is exactly what CASEDIR is for already

File: t/40-script_openqa-clone-custom-git-refspec.t [hunk 2]
```diff
@@ -113,21 +110,4 @@ $expected    = $clone_job . '1169326 _GROUP=0 TEST=my_test@user/repo#my_branch B
...
-test_once $args, $expected_re, "PRODUCTDIR is correct when the source job's PRODUCTDIR is a relative directory";
-
-$ENV{curl_openqa}
-  = qq{echo -e '{"TEST": "my_test", "CASEDIR": "/var/lib/openqa/cache/openqa1-opensuse/tests/opensuse", "PRODUCTDIR": "/var/lib/openqa/cache/openqa1-opensuse/tests/opensuse/products/opensuse"}'; true};
-$dirs
-  = 'CASEDIR=https://github.com/user/repo.git#my/branch PRODUCTDIR=repo/products/opensuse NEEDLES_DIR=/var/lib/openqa/cache/openqa1-opensuse/tests/opensuse/products/opensuse/needles';
-$expected    = $clone_job . '1169326 _GROUP=0 TEST=my_test@user/repo#my/branch BUILD=user/repo#9539 ' . $dirs;
-$expected_re = qr/${expected}/;
-test_once $args, $expected_re, "PRODUCTDIR is correct when the source job's PRODUCTDIR includes specific word";
-
```
Comment (okurz) [inline t/40-script_openqa-clone-custom-git-refspec.t:132]: I think you remove this block because this handles PRODUCTDIR which is now in os-autoinst but I am not sure if by deleting we miss some cases. WDYT?

**Review by okurz — CHANGES_REQUESTED** [review]
As you change the interface in worker code as well as the script openqa-clone-custom-git-refspec I suggest you test (manually) also combinations of different versions of openQA worker code, openqa-clone-custom-git-refspec, os-autoinst. At best all is compatible.

File: t/40-script_openqa-clone-custom-git-refspec.t (hunk 2 above)
Comment (Amrysliu) [reply t/40-script_openqa-clone-custom-git-refspec.t:132]: yes, this block handles PRODUCTDIR. I do not see what we miss except the compatibility.  For your question about compatibility: this old script works with both the new and old openQA worker code and os-autoinst. But the new script only works with the new openQA and os-autoinst.

File: t/24-worker-jobs.t [hunk 3]
```diff
@@ -1253,4 +1255,51 @@ subtest 'Cache setup error handling' => sub {
       'show the error message that is returned by do_asset_caching correctly';
 };
 
+subtest 'symlink DIR to workers working directory' => sub {
```
Comment (Martchus) [inline t/24-worker-jobs.t:1258]: Likely `t/24-worker-engine.t` would have been more fitting since you're actually modifying `isotovideo.pm` and not `Job.pm`. You likely also don't need to deal with the "life-cycle" of a job object here¹.

¹ I'm referring to code like

```
    $job->accept;
    wait_until_job_status_ok($job, 'accepted');
    combined_like { $job->start 
```

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 4]
```diff
@@ -326,11 +340,28 @@ sub engine_workit {
         my $error = locate_local_assets(\%vars, $assetkeys);
         return $error if $error;
     }
+    my $casedir             = OpenQA::Utils::testcasedir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
+    my $productdir          = OpenQA::Utils::productdir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
+    my $relative_casedir    = substr($casedir,          length($vars{PRJDIR}) + 1);
+    my $relative_productdir = substr($productdir,       length($vars{PRJDIR}) + 1);
+    my $target_name         = substr($relative_casedir, 0, index($relative_casedir, '/'));
```
Comment (Martchus) [inline lib/OpenQA/Worker/Engines/isotovideo.pm:347]: On the first look it seems a little bit odd to construct these absolute paths just to make them relative again using `substr` and `length`. Maybe it makes more sense to adjust the 2 utility functions so they can directly return paths relative to the share dir? Or add relative versions of them, e.g. `testcasedir_rel` and `productdir_rel`. The absolute versions could maybe be implemented in terms of the relative ones by simply adding the absolute prefix afterwards.

Note that this code is very sub-optimal anyways because `testcasedir` might return a path which is under `$prjdir/tests` and not under `$prjdir/share`. Your code unconditionally removes the first `length($vars{PRJDIR})` characters which works only in both cases because `tests` and `share` have coincidentally the same number of characters. It also means that the directory the relative paths are relative to depends on the outcome of `testcasedir`. However, I suppose it should always be relative to `sharedir`/`$vars{PRJDIR}`.

Maybe we could also drop the case where we look for tests under `$prjdir/tests` as is complicates things. @okurz What do you think?

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 5]
```diff
@@ -326,11 +340,28 @@ sub engine_workit {
...
+    # if the CASEDIR is a git repository, we do not need to do symlink for CASEDIR.
+    # The code will be git clone in os-autoinst.
+    if ($vars{CASEDIR} !~ /http(s)?\:\/\//) {
+        my $link_casedir_error = _link_testrepo(catdir($vars{PRJDIR}, $target_name) , $pooldir, $target_name);
+        return $link_casedir_error if $link_casedir_error;
+    }
+    if ($vars{NEEDLES_DIR} && !File::Spec->file_name_is_absolute($vars{NEEDLES_DIR}) && $vars{NEEDLES_DIR} !~ /http(s)?\:\/\//)
+    {
+        my $needles_dir          = $productdir . "/needles";
+        my $link_needledir_error = _link_testrepo($needles_dir, $pooldir, $vars{NEEDLES_DIR});
```
Comment (Martchus) [inline lib/OpenQA/Worker/Engines/isotovideo.pm:362]:
```suggestion
        my $link_needledir_error = _link_testrepo("$productdir/needles", $pooldir, $vars{NEEDLES_DIR});
```

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 6]
```diff
@@ -326,11 +340,28 @@ sub engine_workit {
...
+    $vars{CASEDIR}    //= $relative_casedir;
+    $vars{PRODUCTDIR} //= $relative_productdir;
+
+    # if the CASEDIR is a git repository, we do not need to do symlink for CASEDIR.
+    # The code will be git clone in os-autoinst.
+    if ($vars{CASEDIR} !~ /http(s)?\:\/\//) {
+        my $link_casedir_error = _link_testrepo(catdir($vars{PRJDIR}, $target_name) , $pooldir, $target_name);
+        return $link_casedir_error if $link_casedir_error;
+    }
+    if ($vars{NEEDLES_DIR} && !File::Spec->file_name_is_absolute($vars{NEEDLES_DIR}) && $vars{NEEDLES_DIR} !~ /http(s)?\:\/\//)
```
Comment (Martchus) [inline lib/OpenQA/Worker/Engines/isotovideo.pm:359]: Maybe the check for a Git URL should be moved in a separate function so the regex is not repeated.

**Review by Martchus — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 4 above)
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:347]: After committing the code, I did a test that enabled cache service. Using a relative path compared to `$prjdir` or `$share` is not correct in this scenario. For example:
The values in `vars.json` are
```
PRJDIR=/var/lib/openqa/share
CACHEDIRECTORY=/var/lib/openqa/cache
``` 

the $casedir (abosulte path ) is `/var/lib/openqa/cache/localhost/tests/sle`. so we cannot use PRJDIR or CACHEDIRECTORY to get the relative path.
How about keeping the last version modification, just symlink `$casedir` to `path($casedir)->basename`, so we do not need to care whether we enable or disable the cache service, and the `sharedir`.

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 7]
```diff
@@ -326,11 +340,28 @@ sub engine_workit {
...
-
+    $vars{CASEDIR}    //= $relative_casedir;
+    $vars{PRODUCTDIR} //= $relative_productdir;
+
+    # if the CASEDIR is a git repository, we do not need to do symlink for CASEDIR.
+    # The code will be git clone in os-autoinst.
+    if ($vars{CASEDIR} !~ /http(s)?\:\/\//) {
+        my $link_casedir_error = _link_testrepo(catdir($vars{PRJDIR}, $target_name) , $pooldir, $target_name);
+        return $link_casedir_error if $link_casedir_error;
+    }
```
Comment (okurz) [inline lib/OpenQA/Worker/Engines/isotovideo.pm:358]: Did you introduce this just to not create an unused symlink? I would prefer if openQA would as of now not care at all about the "git clone feature" within os-autoinst. That os-autoinst might not use the symlink for needles should not be our concern here

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 7 above)
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: yes, my purpose was that. I removed that `if` block.

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 6 above)
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: According to https://github.com/os-autoinst/openQA/pull/3604#discussion_r544610157,  I removed the `if` block when doing symlink casedir, so there is only one.

File: t/24-worker-engine.t [hunk 8]
```diff
@@ -138,6 +156,7 @@ subtest 'asset caching' => sub {
     ok $job->called('client'), 'client has been asked for parameters when accessing job for caching';
     is $got, undef, 'Assets cached but not tests';
     $testpool_server = 'host1';
+    print $job->client->testpool_server;
```
Comment (kalikiana) [inline t/24-worker-engine.t:159]: This looks like a debug print left over

File: t/24-worker-engine.t [hunk 9]
```diff
@@ -26,21 +26,39 @@ use Test::MockModule;
...
-use Mojo::File 'path';
+use Mojo::File qw(path tempdir);
+use OpenQA::Utils 'testcasedir';
 
 # define fake packages for testing asset caching
 {
     package Test::FakeJob;
     use Mojo::Base -base;
     has id => 42;
+
```
Comment (okurz) [inline t/24-worker-engine.t:37]: unrelated change, please revert

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 10]
```diff
@@ -326,11 +340,24 @@ sub engine_workit {
...
+    my $link_casedir_error = _link_testrepo($casedir, $pooldir, $target_name);
+    return $link_casedir_error if $link_casedir_error;
+
+    if (   $vars{NEEDLES_DIR}
+        && !File::Spec->file_name_is_absolute($vars{NEEDLES_DIR})
+        && $vars{NEEDLES_DIR} !~ /http(s)?\:\/\//)
+    {
+        my $link_needledir_error = _link_testrepo("$productdir/needles", $pooldir, $vars{NEEDLES_DIR});
+        return $link_needledir_error if $link_needledir_error;
+    }
```
[the remark this answers is not in the document]
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:360]: I just think we do not need to do the symlink because we know os-autoinst will clone the needles and reset the NEEDLES_DIR. And the `symlink("$productdir/needles",  "$pool_dir/https://github.com/xxxx")` will failed, so maybe we skip doing symlink is better.

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 11]
```diff
@@ -326,11 +340,24 @@ sub engine_workit {
...
 
     $vars{ASSETDIR}   //= OpenQA::Utils::assetdir();
-    $vars{CASEDIR}    //= OpenQA::Utils::testcasedir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
-    $vars{PRODUCTDIR} //= OpenQA::Utils::productdir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
-
+    $vars{CASEDIR}    //= $target_name;
+    $vars{PRODUCTDIR} //= substr($productdir, rindex($casedir, $target_name));
+
+    my $link_casedir_error = _link_testrepo($casedir, $pooldir, $target_name);
+    return $link_casedir_error if $link_casedir_error;
```
[the remark this answers is not in the document]
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:352]: Actually, I do not have good reasons to handle the error. I just follow the code in `_link_asset` in https://github.com/os-autoinst/openQA/blob/master/lib/OpenQA/Worker/Engines/isotovideo.pm#L172. Although it's too defensive, we could know the fail details. And if `symlink` fails,  the job will fail when running `isotovideo ....`,  I guess we quit the job earlier is not a bad thing.

[the remark this answers is not in the document]
Comment (okurz) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: yes, failing early and hard is a good approach to find out what is actually to be expected :)

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 10 above)
[the remark this answers is not in the document]
Comment (okurz) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: Similar as above: I think you could simply call

```perl
symlink($vars{NEEDLES_DIR}, $pooldir->child('needles')) if $vars{NEEDLES_DIR};
```

and ignore the error cases when a symlink can not be created

[the remark this answers is not in the document]
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: So as the above comments `failing early and hard is a good approach to find out what is actually to be expected`, could we keep the error handling and `http` check here? I prefer to keep it.  WDYT?  @Martchus

[the remark this answers is not in the document]
Comment (Martchus) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: I would also avoid ignoring any error blindly, e.g. a permission error might occur and we shouldn't pretend everything is good in this case.

File: script/openqa-clone-custom-git-refspec [hunk 12]
```diff
@@ -127,7 +127,7 @@ Please try 'curl $json_url' or select another job, e.g. in the same scenario: $h
         local old_needledir
         old_needledir=$(echo "$json_data" | jq -r '.NEEDLES_DIR | select (.!=null)') || throw_json_error "$json_url" "$json_data"
         local needles_dir="${needles_dir:-"$old_needledir"}"
-        needles_dir="${needles_dir:-"$old_productdir/needles"}"
+        needles_dir="${needles_dir:-"needles"}"
```
Comment (okurz) [inline script/openqa-clone-custom-git-refspec:130]: can we keep the script in the old form around a bit longer so that we have a chance of rolling out the os-autoinst and openQA server changes first before rolling out new scripts that only work with newer os-autoinst and server?

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 10 above)
[the remark this answers is not in the document]
Comment (okurz) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: ok, fine. But still I think you should be able to simply do a one-line call to symlink pluse one-line error handling and not need a separate function for that. Something like

```perl
my $err = symlink($vars{NEEDLES_DIR}, $pooldir->child('needles')) if $vars{NEEDLES_DIR} && !-e $pooldir->child('needles');
die 'Could not symlink ' . $vars{NEEDLES_DIR} ' into pool directory' if $err;
```

**Comment by okurz** [comment]
I see that you removed the "acceptance-tests-needed" label. Could you please share your test results in the git commit message or PR description or as updated comment in the PR according to what you see fitting?

File: script/openqa-clone-custom-git-refspec (hunk 12 above)
Comment (Amrysliu) [reply script/openqa-clone-custom-git-refspec:130]: I think we can keep this script in the old form around a bit longer if we keep that check
```
if (   $vars{NEEDLES_DIR}
        && !File::Spec->file_name_is_absolute($vars{NEEDLES_DIR})
        && $vars{NEEDLES_DIR} !~ /http(s)?\:\/\//)
    {
        if (my $error = _link_repo("$productdir/needles", $pooldir, $vars{NEEDLES_DIR})) { return $error }
    }
```

**Comment by Amrysliu** [comment]
> I see that you removed the "acceptance-tests-needed" label. Could you please share your test results in the git commit message or PR description or as updated comment in the PR according to what you see fitting?

Added in the PR description.

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 10 above)
[the remark this answers is not in the document]
Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: I make it a separate function because I could write more details (such as print the debug log message when symlink success) and do more checking (this checking seems to be defensive).  does that way break our coding style? If it does, I do not mind changing my code.  If it does not, I don't think it's necessary to reduce these lines.

File: script/openqa-clone-custom-git-refspec (hunk 12 above)
Comment (okurz) [reply script/openqa-clone-custom-git-refspec:130]: good, let's keep this then and you can change the script back to the old version

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 13]
```diff
@@ -192,6 +192,20 @@ sub _link_asset {
...
+sub _link_repo {
+    my ($source_dir, $pooldir, $target_name) = @_;
+    $pooldir = path($pooldir);
+    my $target = $pooldir->child($target_name);
+    unlink $target if -e $target;
+    return {error => "The source directory $source_dir does not exist"} unless -e $source_dir;
+    eval { symlink($source_dir, $target) or die qq{Cannot create symlink from "$source_dir" to "$target": $!} };
+    if (my $error = $@) {
+        return {error => $error};
+    }
```
Comment (okurz) [inline lib/OpenQA/Worker/Engines/isotovideo.pm:204]: Please make that single line:

```suggestion
    if (my $error = $@) { return {error => $error} }
```

Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm]: fixed

File: script/openqa-clone-custom-git-refspec (hunk 12 above)
Comment (Amrysliu) [reply script/openqa-clone-custom-git-refspec]: I have changed the script back to the old version. Just a comment, cloning a job (run on a remote worker) still has the problem if we keep this script as the old version.

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 14]
```diff
@@ -326,11 +338,22 @@ sub engine_workit {
...
-    $vars{CASEDIR}    //= OpenQA::Utils::testcasedir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
-    $vars{PRODUCTDIR} //= OpenQA::Utils::productdir($vars{DISTRI}, $vars{VERSION}, $shared_cache);
+    $vars{CASEDIR}    //= $target_name;
+    $vars{PRODUCTDIR} //= substr($productdir, rindex($casedir, $target_name));
+
+    if (my $error = _link_repo($casedir, $pooldir, $target_name)) { return $error }
 
+    if (   $vars{NEEDLES_DIR}
+        && !File::Spec->file_name_is_absolute($vars{NEEDLES_DIR})
+        && $vars{NEEDLES_DIR} !~ /http(s)?\:\/\//)
```
Comment (okurz) [inline lib/OpenQA/Worker/Engines/isotovideo.pm:353]: Please add a comment explaining the speciality about "http" here.

Comment (kalikiana) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:353]: This is meant to check if `NEEDLES_DIR` is a relative path to a local folder, isn't it? (it's not absolute, it's not http or https). In that case it might be cleaner to check the opposite. See if this is a valid, relative folder that exists. Otherwise leave it alone.

Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:356]: added the comments

Comment (Amrysliu) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:356]: If I understand correctly, usually, we don't specify NEEDLE_DIRS here, only users need to get needles from URL address (github or gitlab), or users use `openqa-clone-custom-git-refspec`. For now, this script will give an absolute path as a NEEDLE_DIR, such as `NEEDLES_DIR=/var/lib/openqa/cache/openqa.suse.de/sle/products/sle/needles`. We don't need to do symlink in these scenarios. 
In the future, we will change the `NEEDLE_DIR=needles` in `openqa-clone-custom-git-refspec`, we need to do symlink at that time.

Comment (kalikiana) [reply lib/OpenQA/Worker/Engines/isotovideo.pm:356]: That sounds good 👍🏽  My idea was just to look at the filesystem and see if the relative path exists - but maybe we get a better error message the way you implemented it. Thank you for expanding on your motivation.

**Review by kalikiana — APPROVED** [review]

File: t/43-scheduling-and-worker-scalability.t [hunk 15]
```diff
@@ -116,16 +117,22 @@ sub log_jobs {
...
-    VERSION => 'Factory',
+    DISTRI  => $distri,
+    VERSION => $version,
     FLAVOR  => 'tape',
     ARCH    => 'x86_64',
     MACHINE => 'xxx',
 );
 $job_ids{$jobs->create({@job_settings, TEST => "dummy-$_"})->id} = 1 for 1 .. $job_count;
 
+# the casedir must exist after making symlink from casedir to work directory
```
[the remark this answers is not in the document]
Comment (kalikiana) [reply t/43-scheduling-and-worker-scalability.t]: 🤔

**Review by okurz — APPROVED** [review]

**Review by okurz — CHANGES_REQUESTED** [review]
Just some comments regarding the git commit message. Please describe in the git commit message what we/you want to achieve, not just what you change. Describe the reason or the benefit of your changes. Also, there is no variable "POOLDIR" AFAIK so please not write it in upper-case this way.

**Review by okurz — CHANGES_REQUESTED** [review]
s/CAESEDIR/CASEDIR/ in git commit message, rest is fine

**Review by Martchus — COMMENTED** [review]
I guess it is ok now apart from the typo mentioned by @okurz.

**Comment by Amrysliu** [comment]
> s/CAESEDIR/CASEDIR/ in git commit message, rest is fine

@okurz @Martchus Thanks for reviewing. And I am sorry for this stupid issue.

**Comment by coolo** [comment]
As this was deployed on o3, it breaks my use case of having a test distri that is not on the webui. `DISTRI=mine BUILD=test CASEDIR=https://github.com/coolo/os-autoinst-distri-example.git#master` worked fine until yesterday, today it says that `mine` is not part of the rsync directory.

This is the same use case we have for carwos. Please take that into account

**Comment by okurz** [comment]
thanks for the report. Reverted in https://github.com/os-autoinst/openQA/pull/3693
